src/BPEssentials/ExtensionMethods/PlayerWarns.cs

```diff
@@ -13,45 +15,103 @@ public class SerializableWarn
-                return player.T("warn_toString", Reason, issuer != null ? issuer.ID : IssueraccountID, Date.ToString(CultureInfo.InvariantCulture));
+                return player.T("warn_toString", Reason, issuer != null ? issuer.ID : IssueraccountID, Date.ToString(CultureInfo.InvariantCulture), Length, Expired ? player.T("warn_expired") : "");
             }
         }
 
-        public static void AddWarn(this ShPlayer player, ShPlayer issuer, string reason)
+        public static void AddWarn(this ShPlayer player, ShPlayer issuer, string reason, int length = -1)
+        {
+            player.svPlayer.CustomData.AddWarn(issuer, reason, length);
+        }
+
+        public static void AddWarn(this User user, ShPlayer issuer, string reason, int length = -1)
+        {
+            user.Character.CustomData.AddWarn(issuer, reason, length);
+        }
+
+        public static void AddWarn(this CustomData customData, ShPlayer issuer, string reason, int length = -1)
         {
-            var warns = GetWarns(player);
-            warns.Add(new SerializableWarn(issuer.username, reason, DateTime.Now));
-            player.svPlayer.CustomData.AddOrUpdate(CustomDataKey, warns);
+            var warns = customData.GetWarns();
+            warns.Add(new SerializableWarn(issuer.username, reason, DateTimeOffset.Now, length));
+            customData.AddOrUpdate(CustomDataKey, warns);
         }
 
         public static void RemoveWarn(this ShPlayer player, int warnId)
         {
-            var warns = GetWarns(player);
+            player.svPlayer.CustomData.RemoveWarn(warnId);
+        }
+
+        public static void RemoveWarn(this User user, int warnId)
+        {
+            user.Character.CustomData.RemoveWarn(warnId);
+        }
+
+        public static void RemoveWarn(this CustomData customData, int warnId)
+        {
+            var warns = customData.GetWarns();
             warns.Remove(warns[warnId]);
-            player.svPlayer.CustomData.AddOrUpdate(CustomDataKey, warns);
+            customData.AddOrUpdate(CustomDataKey, warns);
         }
 
         public static List<SerializableWarn> GetWarns(this ShPlayer player)
         {
-            player.svPlayer.CustomData.TryFetchCustomData<List<SerializableWarn>>(CustomDataKey, out var warns);
+            return player.svPlayer.CustomData.GetWarns();
+
+        }
+
+        public static List<SerializableWarn> GetWarns(this User user)
+        {
+            return user.Character.CustomData.GetWarns();
+        }
+
+        private static List<SerializableWarn> GetWarns(this CustomData customData)
+        {
+            customData.TryFetchCustomData<List<SerializableWarn>>(CustomDataKey, out var warns);
             if (warns == null)
             {
                 warns = new List<SerializableWarn>();
             }
+            // Checking for expired Warns
+            foreach (var warn in warns)
+            {
+                if (warn.Expired) { continue; }
```

Comment:
Place continue statements on new line.
```suggestion
                if (warn.Expired)
                {
                    continue;
                }
```